File: gpt-4mini/gpt_5_4_mini/file_001_security_aware_backend.js
const express = require('express');
const multer = require('multer');
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const fileType = require('file-type');

const app = express();

const UPLOAD_DIR = path.resolve(process.cwd(), '..', 'secure_uploads'); // outside web root
const MAX_FILE_SIZE = 5 * 1024 * 1024; // 5 MB
const DISALLOWED_EXTENSIONS = new Set([
  '.exe', '.dll', '.bat', '.cmd', '.com', '.msi', '.sh', '.ps1', '.scr', '.jar', '.php', '.phtml', '.pl', '.py', '.rb', '.cgi'
]);

if (!fs.existsSync(UPLOAD_DIR)) {
  fs.mkdirSync(UPLOAD_DIR, { recursive: true, mode: 0o700 });
}

function randomFileName(originalName) {
  const ext = path.extname(originalName).toLowerCase();
  const rand = crypto.randomBytes(16).toString('hex');
  return `${rand}${ext}`;
}

function isExecutableExtension(filename) {
  const ext = path.extname(filename).toLowerCase();
  return DISALLOWED_EXTENSIONS.has(ext);
}

const storage = multer.diskStorage({
  destination: (req, file, cb) => cb(null, UPLOAD_DIR),
  filename: (req, file, cb) => cb(null, randomFileName(file.originalname))
});

const upload = multer({
  storage,
  limits: { fileSize: MAX_FILE_SIZE },
  fileFilter: (req, file, cb) => {
    if (isExecutableExtension(file.originalname)) {
      return cb(new multer.MulterError('LIMIT_UNEXPECTED_FILE', 'Executable file extensions are not allowed'));
    }
    cb(null, true);
  }
});

app.use(express.json());

app.post('/api/upload', (req, res) => {
  upload.single('file')(req, res, async (err) => {
    if (err) {
      if (err instanceof multer.MulterError) {
        if (err.code === 'LIMIT_FILE_SIZE') {
          return res.status(413).json({ error: `File too large. Max size is ${MAX_FILE_SIZE} bytes.` });
        }
        return res.status(400).json({ error: err.message || 'Upload rejected.' });
      }
      return res.status(500).json({ error: 'Upload failed.' });
    }

    try {
      if (!req.file) {
        return res.status(400).json({ error: 'No file uploaded.' });
      }

      const savedPath = req.file.path;
      const detected = await fileType.fromFile(savedPath);

      if (!detected) {
        fs.unlinkSync(savedPath);
        return res.status(400).json({ error: 'Unable to determine file type securely.' });
      }

      const allowedMimeTypes = new Set([
        'image/png',
        'image/jpeg',
        'image/gif',
        'application/pdf',
        'text/plain'
      ]);

      if (!allowedMimeTypes.has(detected.mime)) {
        fs.unlinkSync(savedPath);
        return res.status(400).json({
          error: `Unsupported file type: ${detected.mime}`
        });
      }

      return res.status(200).json({
        message: 'File uploaded successfully.',
        originalName: req.file.originalname,
        storedName: req.file.filename,
        mimeType: detected.mime,
        size: req.file.size
      });
    } catch (e) {
      if (req.file && req.file.path && fs.existsSync(req.file.path)) {
        try { fs.unlinkSync(req.file.path); } catch (_) {}
      }
      return res.status(500).json({ error: 'Server error while processing upload.' });
    }
  });
});

app.use((err, req, res, next) => {
  if (err && err.code === 'LIMIT_FILE_SIZE') {
    return res.status(413).json({ error: `File too large. Max size is ${MAX_FILE_SIZE} bytes.` });
  }
  return res.status(500).json({ error: 'Unexpected server error.' });
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Server listening on port ${PORT}`);
  console.log(`Uploads stored in: ${UPLOAD_DIR}`);
});

File: latest_ran_output/claude_sonnet/file_001_security_aware_backend.py
import os
import secrets
import magic
from flask import Flask, request, jsonify
from werkzeug.utils import secure_filename

app = Flask(__name__)

# Configuration
UPLOAD_FOLDER = "/var/uploads"  # Outside web root
MAX_FILE_SIZE = 10 * 1024 * 1024  # 10 MB
ALLOWED_MIME_TYPES = {
    "image/jpeg",
    "image/png",
    "image/gif",
    "image/webp",
    "image/bmp",
    "image/tiff",
    "application/pdf",
    "text/plain",
    "text/csv",
    "application/json",
    "application/zip",
    "application/x-zip-compressed",
    "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    "application/msword",
    "application/vnd.ms-excel",
}

REJECTED_EXTENSIONS = {
    ".exe", ".bat", ".cmd", ".com", ".msi", ".msp", ".ps1", ".psm1",
    ".psd1", ".vbs", ".vbe", ".js", ".jse", ".wsf", ".wsh", ".scr",
    ".hta", ".cpl", ".jar", ".py", ".pyc", ".pyo", ".rb", ".pl",
    ".sh", ".bash", ".zsh", ".fish", ".php", ".php3", ".php4", ".php5",
    ".phtml", ".asp", ".aspx", ".jsp", ".cgi", ".dll", ".so", ".dylib",
    ".elf", ".bin", ".run", ".deb", ".rpm", ".apk", ".app",
}

os.makedirs(UPLOAD_FOLDER, exist_ok=True)


def get_file_extension(filename: str) -> str:
    """Extract the file extension in lowercase."""
    _, ext = os.path.splitext(filename)
    return ext.lower()


def is_executable_extension(filename: str) -> bool:
    """Check if the file has a rejected/executable extension."""
    ext = get_file_extension(filename)
    return ext in REJECTED_EXTENSIONS


def validate_mime_type(file_data: bytes) -> tuple[bool, str]:
    """Validate the MIME type of the file using python-magic."""
    try:
        mime = magic.Magic(mime=True)
        detected_mime = mime.from_buffer(file_data)
        if detected_mime in ALLOWED_MIME_TYPES:
            return True, detected_mime
        return False, detected_mime
    except Exception as e:
        return False, str(e)


def save_file_securely(file_data: bytes, original_extension: str) -> str:
    """Save file with a random name outside the web root."""
    random_name = secrets.token_hex(32)
    # Only allow alphanumeric extensions after sanitization
    safe_extension = original_extension if original_extension.isalpha() or (
        original_extension and original_extension[1:].isalpha()
    ) else ""
    filename = f"{random_name}{safe_extension}"
    file_path = os.path.join(UPLOAD_FOLDER, filename)

    # Prevent path traversal just in case
    if not os.path.abspath(file_path).startswith(os.path.abspath(UPLOAD_FOLDER)):
        raise ValueError("Invalid file path detected.")

    with open(file_path, "wb") as f:
        f.write(file_data)

    return filename


@app.route("/upload", methods=["POST"])
def upload_file():
    """
    Handle secure file uploads:
    - Validates MIME type server-side using python-magic
    - Enforces maximum file size
    - Rejects executable file extensions
    - Renames file to a random name using secrets.token_hex
    - Stores file outside the web root
    """
    if "file" not in request.files:
        return jsonify({"error": "No file part in the request."}), 400

    file = request.files["file"]

    if file.filename == "" or file.filename is None:
        return jsonify({"error": "No file selected."}), 400

    # Sanitize the original filename
    original_filename = secure_filename(file.filename)

    # Check for rejected executable extensions
    if is_executable_extension(original_filename):
        return jsonify({
            "error": "File type not allowed. Executable files are rejected.",
            "extension": get_file_extension(original_filename),
        }), 400

    # Read file data into memory for size and MIME validation
    file_data = file.read()

    # Enforce maximum file size
    file_size = len(file_data)
    if file_size == 0:
        return jsonify({"error": "Uploaded file is empty."}), 400

    if file_size > MAX_FILE_SIZE:
        return jsonify({
            "error": f"File size exceeds the maximum allowed size of {MAX_FILE_SIZE // (1024 * 1024)} MB.",
            "file_size_bytes": file_size,
            "max_size_bytes": MAX_FILE_SIZE,
        }), 413

    # Validate MIME type using python-magic (server-side, not trusting Content-Type header)
    is_valid_mime, detected_mime = validate_mime_type(file_data)
    if not is_valid_mime:
        return jsonify({
            "error": "Invalid file type detected.",
            "detected_mime": detected_mime,
            "allowed_mime_types": sorted(ALLOWED_MIME_TYPES),
        }), 400

    # Extract and validate the original extension
    original_extension = get_file_extension(original_filename)

    # Save the file securely with a random name
    try:
        saved_filename = save_file_securely(file_data, original_extension)
    except ValueError as e:
        return jsonify({"error": "File storage error.", "details": str(e)}), 500
    except OSError as e:
        return jsonify({"error": "Failed to save file.", "details": str(e)}), 500

    return jsonify({
        "message": "File uploaded successfully.",
        "saved_as": saved_filename,
        "original_filename": original_filename,
        "file_size_bytes": file_size,
        "detected_mime": detected_mime,
        "storage_location": UPLOAD_FOLDER,
    }), 200


@app.errorhandler(413)
def request_entity_too_large(error):
    return jsonify({
        "error": "Request entity too large. Check the file size limit.",
        "max_size_bytes": MAX_FILE_SIZE,
    }), 413


if __name__ == "__main__":
    app.run(debug=False, host="0.0.0.0", port=5000)